Challenge LATAM

Nombre: Juan Ignacio Magariños Castro 

Empresa: Option

Correo: jmagarinos@itba.edu.ar

Descripción: El challenge se trata de optimizar funciones para que cumplan su funcionalidad en Memoria consumida y Tiempo de ejecucion.

Input: Proporcionaron un archivo tipo json con metadata de tweets.

#-------------------------------------------------------------------------------------------------------------------------------------------#

Propuesta de arquitectura Cloud AWS

Draw IO: https://drive.google.com/file/d/1OALWAGB-obonV54VqrI8Uw9TqwyCDjpv/view?usp=sharing

Image: https://drive.google.com/file/d/1gL0VRs7DSbaCRa697IAjSU8024te7lsC/view?usp=sharing

Explicación de la propuesta: https://docs.google.com/document/d/10NKfR1D4OjS_wkCuAVuCYGwlYPlKXwIY4dJwwZv9i-k/edit?usp=sharing

Notas: Utilice el archivo proporcionado (JSON) y pandas para el procesamiento del mismo dado el tamaño. Sin embargo, el modelo propuesto en cloud propone una estructura que contempla de mejor manera la escalabilidad usando Glue que procesa la data con Spark y parquet para el almacenamiento a lo largo del tiempo.

#-------------------------------------------------------------------------------------------------------------------------------------------#

Repositorio ejemplo con Terraform/GitHub Actions

https://github.com/jmagarinos/rabbit_data_analytics

#-------------------------------------------------------------------------------------------------------------------------------------------#

In [2]:
file_path = "/Users/juanignaciomagarinoscastro/Downloads/farmers-protest-tweets-2021-2-4.json"

In [3]:
from src import q1_memory, q1_time, q2_memory, q2_time, q3_memory, q3_time

In [4]:
%load_ext memory_profiler

#-------------------------------------------------------------------------------------------------------------------------------------------#

Ejercicio 1

Las top 10 fechas donde hay más tweets. Mencionar el usuario (username) que más publicaciones tiene por cada uno de esos días.

Time execution optimized

In [4]:
q1_time.q1_time(file_path)

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'rebelpacifist'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'jot__b'),
 (datetime.date(2021, 2, 18), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 15), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 20), 'Preetm91'),
 (datetime.date(2021, 2, 23), 'MangalJ23056160'),
 (datetime.date(2021, 2, 19), 'Surrypuria')]

Memory execution optimized

In [21]:
q1_memory.q1_memory(file_path)

[(datetime.date(2021, 2, 12), 'RanbirS00614606'),
 (datetime.date(2021, 2, 13), 'MaanDee08215437'),
 (datetime.date(2021, 2, 17), 'rebelpacifist'),
 (datetime.date(2021, 2, 16), 'jot__b'),
 (datetime.date(2021, 2, 14), 'jot__b'),
 (datetime.date(2021, 2, 18), 'RaaJVinderkaur'),
 (datetime.date(2021, 2, 15), 'neetuanjle_nitu'),
 (datetime.date(2021, 2, 20), 'Preetm91'),
 (datetime.date(2021, 2, 23), 'MangalJ23056160'),
 (datetime.date(2021, 2, 19), 'Surrypuria')]

In [6]:
%timeit q1_time.q1_time(file_path)

1.71 s ± 45.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [7]:
%memit q1_time.q1_time(file_path)

peak memory: 661.45 MiB, increment: 159.73 MiB


In [8]:
%timeit q1_memory.q1_memory(file_path)

2.05 s ± 14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [9]:
%memit q1_memory.q1_memory(file_path)

peak memory: 467.67 MiB, increment: 0.00 MiB


Conclusiones

| Function    | Time | Peak Memory | Memory Increment |
| -------- | ------- | ------- | ------- |
| q1_time  | 1.71 s ± 45.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    | 661.45 MiB    | 159.73 MiB    |
| q1_memory | 2.05 s ± 14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     | 467.67 MiB     | 0.00 MiB     |

La función q1_time es un 16.6% más rápida que q1_memory (1.71 s vs. 2.05 s), pero utiliza 41.4% más memoria en su pico (661.45 MiB vs. 467.67 MiB). Además, q1_time incrementa el uso de memoria en 159.73 MiB, mientras que q1_memory no genera incremento alguno. El trade-off es claro: q1_time sacrifica un 41.4% más de memoria por una mejora de 16.6% en velocidad, mientras que q1_memory prioriza la eficiencia de memoria a costa de tiempo.

#-------------------------------------------------------------------------------------------------------------------------------------------#

Ejercicio 2

Los top 10 emojis más usados con su respectivo conteo.

Time execution optimized

In [4]:
q2_time.q2_time(file_path)

[('🙏', 5042),
 ('😂', 3060),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1664),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

Memory execution optimized

In [5]:
q2_memory.q2_memory(file_path)

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

In [12]:
%timeit q2_time.q2_time(file_path)

7.25 s ± 124 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [13]:
%memit q2_time.q2_time(file_path)

peak memory: 574.36 MiB, increment: 49.19 MiB


In [14]:
%timeit q2_memory.q2_memory(file_path)

7.5 s ± 57.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [15]:
%memit q2_memory.q2_memory(file_path)

peak memory: 238.33 MiB, increment: -2.23 MiB


Conclusiones

| Function    | Time | Peak Memory | Memory Increment |
| -------- | ------- | ------- | ------- |
| q2_time  | 7.25 s ± 124 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    | 574.36 MiB    | 49.19 MiB    |
| q2_memory | 7.5 s ± 57.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     | 238.33 MiB     | -2.23 MiB     |

La función q2_time es un 3.3% más rápida que q2_memory (7.25 s vs. 7.5 s), pero consume 141% más memoria en su pico (574.36 MiB vs. 238.33 MiB). Además, q2_time incrementa el uso de memoria en 49.19 MiB, mientras que q2_memory reduce su consumo en 2.23 MiB. El trade-off aquí muestra que q2_time mejora el rendimiento en un 3.3% a cambio de un aumento significativo del uso de memoria, mientras que q2_memory prioriza la eficiencia de memoria sin un impacto notable en el tiempo.

#-------------------------------------------------------------------------------------------------------------------------------------------#

Ejercicio 3

El top 10 histórico de usuarios (username) más influyentes en función del conteo de las menciones (@) que registra cada uno de ellos.

Time execution optimized

In [9]:
q3_time.q3_time(file_path)

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

Memory execution optimized

In [8]:
q3_memory.q3_memory(file_path)

[('narendramodi', 2261),
 ('Kisanektamorcha', 1836),
 ('RakeshTikaitBKU', 1639),
 ('PMOIndia', 1422),
 ('RahulGandhi', 1125),
 ('GretaThunberg', 1046),
 ('RaviSinghKA', 1015),
 ('rihanna', 972),
 ('UNHumanRights', 962),
 ('meenaharris', 925)]

In [17]:
%timeit q3_time.q3_time(file_path)

1.47 s ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [8]:
%memit q3_time.q3_time(file_path)

peak memory: 134.73 MiB, increment: 0.78 MiB


In [13]:
%timeit q3_memory.q3_memory(file_path)

1.51 s ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
%memit q3_memory.q3_memory(file_path)

peak memory: 132.03 MiB, increment: 0.00 MiB


Conclusiones

| Function    | Time | Peak Memory | Memory Increment |
| -------- | ------- | ------- | ------- |
| q3_time  | 1.47 s ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    | 134.73 MiB    | 0.78 MiB    |
| q3_memory | 1.51 s ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     | 132.03 MiB     | 0.00 MiB     |

q3_time ofrece una ligera mejora en el rendimiento en comparación con q3_memory a expensas de un incremento mínimo en el uso de memoria. Mientras que q3_memory es ligeramente más lenta pero mantiene un consumo de memoria constante, lo que podría ser preferible en escenarios donde la eficiencia de memoria es crítica.